File: tests/importer/onnx/test_squeeze.py
# Copyright 2019-2021 Canaan Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# pylint: disable=invalid-name, unused-argument, import-outside-toplevel
import pytest
import os
import torch
import numpy as np
import sys
import test_util

def _make_module(in_shape, out_channel, kernel_size, dim):

    class SqueezeModule(torch.nn.Module):
        def __init__(self):
            super(SqueezeModule, self).__init__()
            self.conv2d = torch.nn.Conv2d(in_shape[1], out_channel, kernel_size)

        def forward(self, x):
            x = self.conv2d(x)
            x = torch.squeeze(x)

            # There is something wrong when converting pytorch into onnx. Use tf.squeeze(x, dim) instead.
            # x = torch.squeeze(x, dim)
            return x

    return SqueezeModule()

in_shapes = [
    [1, 4, 60, 72],
    [1, 3, 224, 224]
]

out_channels = [
    1,
    3,
]

kernel_sizes = [
    1,
    3,
]

axes = [
    0,
    1,
    2,
    3
]

@pytest.mark.parametrize('in_shape', in_shapes)
@pytest.mark.parametrize('out_channel', out_channels)
@pytest.mark.parametrize('kernel_size', kernel_sizes)
@pytest.mark.parametrize('axis', axes)
def test_squeeze(in_shape, out_channel, kernel_size, axis, request):
    out_shape = [in_shape[0], out_channel, in_shape[2] - kernel_size + 1, in_shape[3] - kernel_size + 1]
    dim = axis if out_shape[axis] == 1 else None
    module = _make_module(in_shape, out_channel, kernel_size, dim)

    # test_util.test_onnx_module(request.node.name, module, in_shape, ['cpu', 'k210', 'k510'])
    test_util.test_onnx_module(request.node.name, module, in_shape, ['k510'])

if __name__ == "__main__":
    pytest.main(['-vv', 'test_squeeze.py'])

File: tests/kernels/test_squeeze.cpp
/* Copyright 2019-2021 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "kernel_test.h"
#include <gtest/gtest.h>
#include <iostream>
#include <nncase/kernels/stackvm/tensor_ops.h>
#include <nncase/runtime/datatypes.h>
#include <nncase/runtime/runtime_tensor.h>
#include <nncase/runtime/simple_types.h>
#include <nncase/runtime/stackvm/opcode.h>
#include <ortki/operators.h>

using namespace nncase;
using namespace nncase::runtime;
using namespace ortki;

class SqueezeTest
    : public KernelTest,
      public ::testing::TestWithParam<std::tuple<nncase::typecode_t, dims_t>> {
  public:
    void SetUp() override {
        auto &&[typecode, l_shape] = GetParam();

        input =
            hrt::create(typecode, l_shape, host_runtime_tensor::pool_cpu_only)
                .expect("create tensor failed");
        init_tensor(input);
    }

    void TearDown() override {}

  protected:
    runtime_tensor input;
};

INSTANTIATE_TEST_SUITE_P(
    Squeeze, SqueezeTest,
    testing::Combine(testing::Values(dt_float32, dt_int32, dt_int16, dt_float64,
                                     dt_int8, dt_uint8, dt_uint16, dt_uint32,
                                     dt_uint64, dt_int64, dt_bfloat16,
                                     dt_float16, dt_boolean),
                     testing::Values(dims_t{1, 3, 1, 16}, dims_t{1, 3, 1, 1})));

TEST_P(SqueezeTest, Squeeze) {
    auto l_ort = runtime_tensor_2_ort_tensor(input);

    // expected
    size_t size = 0;
    int64_t axes_array[] = {0, 2};
    auto axes = hrt::create(dt_int64, {2},
                            {reinterpret_cast<gsl::byte *>(axes_array),
                             sizeof(axes_array)},
                            true, host_runtime_tensor::pool_cpu_only)
                    .expect("create tensor failed");
    auto output_ort = ortki_Squeeze(l_ort, runtime_tensor_2_ort_tensor(axes));
    void *ptr_ort = tensor_buffer(output_ort, &size);
    dims_t shape(tensor_rank(output_ort));
    tensor_shape(output_ort, reinterpret_cast<int64_t *>(shape.data()));
    auto expected = hrt::create(input.datatype(), shape,
                                {reinterpret_cast<gsl::byte *>(ptr_ort), size},
                                true, host_runtime_tensor::pool_cpu_only)
                        .expect("create tensor failed");

    // actual
    auto output = kernels::stackvm::squeeze(input.impl(), axes.impl())
                      .expect("squeeze failed");
    runtime_tensor actual(output.as<tensor>().expect("as tensor failed"));

    bool result = is_same_tensor(expected, actual) ||
                  cosine_similarity_tensor(expected, actual);

    if (!result) {
        std::cout << "actual ";
        print_runtime_tensor(actual);
        std::cout << "expected ";
        print_runtime_tensor(expected);
    }

    // compare
    EXPECT_TRUE(result);
}

int main(int argc, char *argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}